File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/creditcard/enrollment/projection/isproductactive/IsProductActiveProjectionHandler.java
package cloud.ambar.creditcard.enrollment.projection.isproductactive;

import cloud.ambar.common.event.Event;
import cloud.ambar.common.projection.ProjectionHandler;
import cloud.ambar.creditcard.product.event.ProductActivated;
import cloud.ambar.creditcard.product.event.ProductDeactivated;
import cloud.ambar.creditcard.product.event.ProductDefined;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.RequestScope;

@Service
@RequestScope
@RequiredArgsConstructor
public class IsProductActiveProjectionHandler extends ProjectionHandler {
    private final ProductActiveStatusRepository productActiveStatusRepository;

    public void project(final Event event) {
        if (event instanceof ProductDefined) {
            productActiveStatusRepository.save(ProductActiveStatus.builder()
                    .id(event.getAggregateId())
                    .active(false)
                    .build());
        } else if (event instanceof ProductActivated) {
            ProductActiveStatus productActiveStatus = productActiveStatusRepository.findOneById(event.getAggregateId()).orElseThrow();
            productActiveStatus.setActive(true);
            productActiveStatusRepository.save(productActiveStatus);
        } else if (event instanceof ProductDeactivated) {
            ProductActiveStatus productActiveStatus = productActiveStatusRepository.findOneById(event.getAggregateId()).orElseThrow();
            productActiveStatus.setActive(false);
            productActiveStatusRepository.save(productActiveStatus);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Projection/IsProductActive/IsProductActiveProjectionHandler.cs
using CreditCardEnrollment.Common.Projection;
using CreditCardEnrollment.CreditCard.Product.Event;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.IsProductActive;

public class IsProductActiveProjectionHandler : ProjectionHandler {
    private readonly ProductActiveStatusRepository _productActiveStatusRepository;

    public IsProductActiveProjectionHandler(ProductActiveStatusRepository productActiveStatusRepository) {
        _productActiveStatusRepository = productActiveStatusRepository;
    }

    public override void Project(Common.Event.Event @event) {
        switch (@event) {
            case ProductDefined:
                _productActiveStatusRepository.Save(new ProductActiveStatus {
                    Id = @event.AggregateId,
                    Active = false
                });
                break;

            case ProductActivated:
                var activatedStatus = _productActiveStatusRepository.FindOneById(@event.AggregateId) ?? throw new InvalidOperationException();
                activatedStatus.Active = true;
                _productActiveStatusRepository.Save(activatedStatus);
                break;

            case ProductDeactivated:
                var deactivatedStatus = _productActiveStatusRepository.FindOneById(@event.AggregateId) ?? throw new InvalidOperationException();
                deactivatedStatus.Active = false;
                _productActiveStatusRepository.Save(deactivatedStatus);
                break;
        }
    }
}